DAT-409 - Add Socrata harvester

madelinekosse opened the pull request:
Adds a harvester for [Socrata Open Data API discovery endpoints](https://socratadiscovery.docs.apiary.io/#), such as the one offered by [Open Data Camden](https://opendata.camden.gov.uk).

It can be configured on the new harvester form by:
- Entering a URL - so far I have only tried https://opendata.camden.gov.uk as that's all we need for this ticket
- Selecting SODA as the source type
- Adding 'app_token' to the configuration, with an application token associated with your account on the source site. We have one of these for production use (see Sven for info), but for test/development it is free and easy to create one yourself.

Socrata APIs don't have a concept of "dataset" vs "resource" - the endpoint provides a list of resources, which may have relationships to other resources. I have ignored these relationships and created one dataset per resource for now

Files are not downloaded and stored in CKAN, as this would just be a waste of space and another thing to worry about synchronising - we just link to the resource on the source.

Not all resources in the Camden source catalogue have a link to an actual file associated with them. Additionally, some have links that 404 - another GIGO situation. I have decided to treat both in the same way: if we can't get a 200 response from the resource download endpoint, we use the link to the HTML page on the source website where users can view all available information.

There's no guaruntee that this will work for other Socrata APIs, as not all of them will have a web interface that works like this, but it works for the Camden Open Data site and we can cross other bridges when we get to them.

The organisations are given only by name in the source API, so we generate IDs for them based on the name. This means that we see different organisations for different spellings of the same thing, such as "London Borough of Camden Council" and "London Borough of Camden Counil". This is a garbage-in, garbage-out situation, about which I am not inclined to do anything. Some source datasets do not have an associated organisation and will be attributed to whatever organisation was given as default in the harvester. If the organisations generated from this system are too messy, we can just omit the 'remote_orgs' key from the config and use a default organisation for all of them.

madelinekosse committed:
DAT-409 - Add Socrata harvester

Adds a harvester for [Socrata Open Data API discovery
endpoints](https://socratadiscovery.docs.apiary.io/#), such as
the one offered by [Open Data Camden](https://opendata.camden.gov.uk).

It can be configured on the new harvester form by:
- Entering a URL - so far I have only tried https://opendata.camden.gov.uk
as that's all we need for this ticket
- Selecting SODA as the source type
- Adding 'app_token' to the configuration, with an application token
  associated with your account on the source site. We have one of these
for production use (see Sven for info), but for test/development it is
free and easy to create one yourself.

Socrata APIs don't have a concept of "dataset" vs "resource" - the
endpoint provides a list of resources, which may have relationships to
other resources. I have ignored these relationships and created one dataset
per resource for now

Files are not downloaded and stored in CKAN, as this would just be a
waste of space and another thing to worry about synchronising - we
just link to the resource on the source.

Not all resources in the Camden source catalogue have a link to an actual
file associated with them. Additionally, some have links that 404 -
another GIGO situation. I have decided to treat both in the same way: if
we can't get a 200 response from the resource download endpoint, we use
the link to the HTML page on the source website where users can view all
available information.

There's no guaruntee that this will work for other Socrata APIs, as not
all of them will have a web interface that works like this, but it works
for the Camden Open Data site and we can cross other bridges when we get to them.

The organisations are given only by name in the source API, so we
generate IDs for them based on the name. This means that we see
different organisations for different spellings of the same thing, such
as "London Borough of Camden Council" and "London Borough of Camden
Counil". This is a garbage-in, garbage-out situation, about which I am
not inclined to do anything. Some source datasets do not have an
associated organisation and will be attributed to whatever organisation
was given as default in the harvester. If the organisations generated
from this system are too messy, we can just omit the 'remote_orgs' key
from the config and use a default organisation for all of them.

kjoshi reviewed (commented):
This looks good to me!

There are a couple of things that some of the other harvesters do that I think might also need to be done in this one.

One is where we check whether any of the previously-harvested datasets have been deleted upstream and then delete them locally if necessary.

And one is where we make sure we pull across any package_extras from a previously-harvested version of the dataset, so we don't lose the `data_quality` etc. stuff when we re-harvest.

Some of this common stuff was getting spread around different harvesters so in my [PR for the Redbridge Harvester](https://github.com/GreaterLondonAuthority/dfl-ckanext-havester/pull/8) I've moved some of it into the `util` module so we can try and remember that it needs to be run on most harvesters.

If it makes sense to you then feel free to copy across the util functions to this PR and merge this one first, then I can deal with the merge conflicts when I merge the Redbridge harvester. 
(This PR has been open so long that it doesn't seem right to make you have to sort out merge conflicts, just because I've been slow to look at it properly!)

madelinekosse committed:
DAT-409 - Delete datasets no longer present in source

Check all datasets previously imported by this harvester and compare the IDs to
the list returned from the source so we can delete those that are no longer
present. Since the harvester is now aware of the datasets currently in CKAN, I
have used this information to check which ones are new so we only check for
changes when a dataset already exists and only add the organisation and extras
when it is new.

madelinekosse committed:
DAT-409 - Keep extras when updating dataset

When a dataset is updated by the harvester, merging the current and
harvested records overwrites the "extras" field, losing the data quality
and boost factor. This change adds the current values of these fields to
the record before saving the update by using the common util function. I
also refactored the part that adds default values to new dataset records
to use a common util function.

Commit:
Bugfix in Redbridge harvester, small update to utils
